libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,551 @@
+			break;
+		commonPrefix.emplace_back(slot1);
+	}
+
+	Stack stack1Tail = _stack1 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+	Stack stack2Tail = _stack2 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+
+	if (stack1Tail.empty())
+		return commonPrefix + compressStack(stack2Tail);
+	if (stack2Tail.empty())
+		return commonPrefix + compressStack(stack1Tail);
+
+	Stack candidate;
+	for (auto slot: stack1Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	for (auto slot: stack2Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	cxx20::erase_if(candidate, [](StackSlot const& slot) {
+		return holds_alternative<LiteralSlot>(slot) || holds_alternative<FunctionCallReturnLabelSlot>(slot);
+	});
+
+	std::map<size_t, Stack> sortedCandidates;
+
+	auto evaluate = [&](Stack const& _candidate) -> size_t {
+		size_t numOps = 0;
+		Stack testStack = _candidate;
+		auto swap = [&](unsigned _swapDepth) { ++numOps; if (_swapDepth > 16) numOps += 1000; };
+		auto dupOrPush = [&](StackSlot const& _slot)
+		{
+			if (canBeFreelyGenerated(_slot))
+				return;
+			auto depth = util::findOffset(ranges::concat_view(commonPrefix, testStack) | ranges::views::reverse, _slot);
+			if (depth && *depth >= 16)
+				numOps += 1000;
+		};
+		createStackLayout(testStack, stack1Tail, swap, dupOrPush, [&](){} );
+		testStack = _candidate;
+		createStackLayout(testStack, stack2Tail, swap, dupOrPush, [&](){});
+		return numOps;
+	};
+
+	// See https://en.wikipedia.org/wiki/Heap's_algorithm
+	size_t n = candidate.size();
+	sortedCandidates.insert(std::make_pair(evaluate(candidate), candidate));
+	std::vector<size_t> c(n, 0);
+	size_t i = 1;
+	while (i < n)
+	{
+		if (c[i] < i)
+		{
+			if (i & 1)
+				std::swap(candidate.front(), candidate[i]);
+			else
+				std::swap(candidate[c[i]], candidate[i]);
+			sortedCandidates.insert(std::make_pair(evaluate(candidate), candidate));
+			++c[i];
+			// Note that for a proper implementation of the Heap algorithm this would need to revert back to ``i = 1.``
+			// However, the incorrect implementation produces decent result and the proper version would have n!
+			// complexity and is thereby not feasible.
+			++i;
+		}
+		else
+		{
+			c[i] = 0;
+			++i;
+		}
+	}
+
+	return commonPrefix + sortedCandidates.begin()->second;
+}
+
+Stack StackLayoutGenerator::compressStack(Stack _stack)
+{
+	optional<size_t> firstDupOffset;
+	do
+	{
+		if (firstDupOffset)
+		{
+			if (_stack.size() - *firstDupOffset - 1 > 1)
+				std::swap(_stack.at(*firstDupOffset + 1), _stack.back());
+			std::swap(_stack.at(*firstDupOffset), _stack.back());
+			_stack.pop_back();
+			firstDupOffset.reset();
+		}
+		for (auto&& [offset, slot]: _stack | ranges::views::enumerate)
+			if (canBeFreelyGenerated(slot))
+				firstDupOffset = offset;
+			else if (auto dupOffset = util::findOffset(_stack | ranges::views::take(offset), slot))
+				if (_stack.size() - *dupOffset <= 16)
```

Comment:
```suggestion
				if (_stack.size() - offset - *dupOffset <= 16)
```
or something? (because `dupOffset` is computed after `views::take(offset)`

Reply:
``views::take(offset)`` just takes the first ``offset`` elements of the range - so it won't affect the indices.

Reply:
Still don't understand, but if you say so :)

Reply:
``dupOffset`` is always relative to the stack bottom - and whether I stop after ``offset`` slots or not does not change anything about that, I'm still relative to the stack bottom and ``_stack.size() - *dupOffset`` is still the distance from stack top. I generally try to use ``offset`` for everything relative to the bottom and ``depth`` for anything relative to the top.

Reply:
Looking at the whole thing, though, it's of course wrong to look in ``_stack | ranges::views::take(offset)`` in the first place :-).

Reply:
Let me have a closer look and fix it.

Reply:
(not exactly "wrong", but it can potentially compress more slots than it does right now, if there's multiple copies on stack)

Reply:
It actually wasn't wrong, it was just a bit counterintuitive to have the loop start from deepest slot and then not breaking on the first match. I now reversed the order, should be easier to read that way.